### ISSUE TITLE
taxon change obs.taxon update mop up

### ISSUE THREAD

**Issue by loarie** [inaturalist/inaturalist#3855]
During large taxon changes, a small percentage of obs don't update their obs.taxon after IDs are altered
e.g. this obs https://www.inaturalist.org/observations/52975355, following a taxon change appeared like this:
<img width="707" alt="Screenshot 2023-09-01 at 3 38 16 PM" src="https://github.com/inaturalist/inaturalist/assets/598026/7e1f8450-4da2-457b-b40d-911df23e80a8">

loarie spent some time trying to reproduce this in staging by monitoring parts of 
https://github.com/inaturalist/inaturalist/blob/main/app/models/taxon_change.rb#L217
and couldn't, he suspects this only happens in production

As a band-aid, this ticket is to add code to mop up any obs that didn't update properly. Aside from opting out of the community ID, there should be no obs with obs.taxon's matching the input taxon following a taxon_change that inactivates the input (e.g. everything except a taxon_split with the input as an output). This code finds such obs where the obs.taxon didn't update
```
input_taxon_id = taxon_change.input_taxon
opt_out = Observation.
 joins("JOIN preferences p ON p.owner_type = 'User' AND p.owner_id = observations.user_id").
 where("taxon_id = ? AND (p.name = 'community_taxa' AND p.value = 'f')", input_taxon_id).pluck(:id)
o = Observation.where("taxon_id = ? AND id NOT IN (?)", input_taxon_id, opt_out)
```
^^ doesn't exclude obs that didn't fail, but didn't update because the only IDer opted out of automatic_taxonomic_changes which may be hard to query for.

**Comment by loarie** [comment]
to manually mop up the output taxa e.g. here https://www.inaturalist.org/taxon_changes/123781
I'm doing this

```
Observation.where(taxon_id: 1455248).find_in_batches(batch_size: 100) do |batch|
  batch.each do |observation|
    observation.set_community_taxon
    observation.save!
    Identification.update_categories_for_observation( observation )
  end
end
```

**Comment by loarie** [comment]
note to myself that IDs from blocked users also don't get updated

**Comment by loarie** [comment]
@pleary @tiwane re: the large recent treefrog merge that I ran in a rails console screen
https://www.inaturalist.org/taxon_changes/138983

I ran it and got this error on an obs that seemed fine, then I reran and got the same thing on another obs then it went all the way through (I'm not sure if this issue is because its a TaxonMerge which is a bit unusual or not)
```
3.0.4 :017 > tc.commit_records
[INFO 2024-02-06 15:07:54 -0800] <TaxonMerge 138983>: starting commit_records
[INFO 2024-02-06 15:07:54 -0800] <TaxonMerge 138983>: committing identifications
/home/inaturalist/deployment/production/shared/bundle/ruby/3.0.0/gems/activerecord-6.1.7.5/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `exec_params': PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_identifications_on_user_id_and_observation_id" (ActiveRecord::RecordNotUnique)
DETAIL:  Key (user_id, observation_id)=(38280, 1378483) already exists.
/home/inaturalist/deployment/production/shared/bundle/ruby/3.0.0/gems/activerecord-6.1.7.5/lib/active_record/connection_adapters/postgresql_adapter.rb:672:in `exec_params': ERROR:  duplicate key value violates unique constraint "index_identifications_on_user_id_and_observation_id" (PG::UniqueViolation)
DETAIL:  Key (user_id, observation_id)=(38280, 1378483) already exists.

/home/inaturalist/deployment/production/releases/20240203015050/app/models/update_action.rb:304:in `block in arel_attributes_to_es_filters': undefined method `id' for main:Object (NoMethodError)
```

I think because of those errors midway through, there were corrupted records which I mopped up by doing this:
```
tc.input_taxa.each do |t|
  taxon_id = t.id
  Observation.where(taxon_id: taxon_id).find_in_batches(batch_size: 1000) do |group|
    group.each do |observation|
      observation.save!
    end
  end
end
```

**Comment by pleary** [comment]
I merged a branch that adds a a method at the end of `commit_records` that runs these cleanup steps. I also added some error catching that should hopefully catch some of the errors you listed and retry processing if any are encountered. Since this ticket is for the cleanup steps, I'll go ahead and close this now. If we find we need additional clean up steps, or these aren't working properly, this can be reopened with requirements. I'd recommend creating new issues for further issues with the actual taxon change processing
